File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/label/TranslatableLabelProvider.java
package dev.obscuria.obscure_tooltips.client.tooltip.label;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.obscure_tooltips.client.component.TooltipComponent;
import net.minecraft.client.resources.I18n;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.text.TextFormatting;

@Desugar
public record TranslatableLabelProvider(String key) implements LabelProvider {

    public static TranslatableLabelProvider fromJson(JsonObject json) {
        return new TranslatableLabelProvider(JsonUtils.getString(json, "key"));
    }

    @Override
    public TooltipComponent create(ItemStack stack) {
        return TooltipComponent.create(TextFormatting.GRAY + I18n.format(key));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/label/TranslatableLabelProvider.kt
package dev.obscuria.tooltips.client.label

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.ChatFormatting
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack

class TranslatableLabelProvider(
    val key: String
) : ILabelProvider {

    override fun codec(): Codec<TranslatableLabelProvider> = CODEC

    override fun bake(stack: ItemStack): ClientTooltipComponent {
        val component = Component.translatable(key).withStyle(ChatFormatting.GRAY)
        return ClientTooltipComponent.create(component.visualOrderText)
    }

    companion object {

        val CODEC: Codec<TranslatableLabelProvider> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.STRING
                    .fieldOf("key")
                    .forGetter(TranslatableLabelProvider::key)
            ).apply(codec, ::TranslatableLabelProvider)
        }
    }
}